### PR TITLE
only update batch norm averages during training

### PR DESCRIPTION
**Pull request by mikowals**
according to [#1724](https://github.com/tensorflow/tensorflow/issues/1724), a node that is meant to only run in one state of tf.cond needs to be defined inside that conditional function.

### CODE REVIEW

**Comment by mikowals** [comment]
I signed it!

**Comment by ilblackdragon** [comment]
@tensorflow-jenkins test this please.

**Comment by mikowals** [comment]
Am I reading the test failures correctly that my change in the true branch is now causing an error to [be thrown by the false branch](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/control_flow_ops.py#L1181)?  The specific error is because `ema.average()` will [return `None`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/moving_averages.py#L316) until `ema.apply()` has been used and `None` does not have a `name` property?

**Comment by yuanbyu** [comment]
The false branch is `lambda: (ema_mean, ema_var)`.  Both ema_mean and ema_var must not be None. Otherwise the computation is meaningless.  The error message should have been better.

**Comment by mikowals** [comment]
I have added a check for None values in the false branch.  Should there also be a warning message when an ExponentialMovingAverage().average() is None, since accessing a, average before any values have been applied seems like a mistake.

File: tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py [hunk 1]
```diff
@@ -45,20 +45,32 @@ def batch_normalize(tensor_in, epsilon=1e-5, convnet=False, decay=0.9,
...
-        ema = moving_averages.ExponentialMovingAverage(decay=decay)
+        moving_mean = vs.get_variable("moving_mean", [shape[-1]],
+                               initializer=init_ops.constant_initializer(0.),
+                               trainable=False)
+        moving_var = vs.get_variable("moving_var", [shape[-1]],
+                               initializer=init_ops.constant_initializer(1.), 
+                               trainable=False)
         if convnet:
-            assign_mean, assign_var = nn.moments(tensor_in, [0, 1, 2])
+            assign_mean, assign_var = nn.moments(tensor_in,  [0, 1, 2])
```
Comment (ilblackdragon) [inline tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py]: remove extra space.

File: tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py [hunk 2]
```diff
@@ -45,20 +45,32 @@ def batch_normalize(tensor_in, epsilon=1e-5, convnet=False, decay=0.9,
...
-            with ops.control_dependencies([ema_assign_op]):
+            update_moving_mean = moving_averages.assign_moving_average(moving_mean, assign_mean, decay)
+            update_moving_var = moving_averages.assign_moving_average(moving_var, assign_var, decay)
+            with ops.control_dependencies([update_moving_mean, update_moving_var]):
                 return array_ops_.identity(assign_mean), array_ops_.identity(assign_var)
+        
         is_training = array_ops_.squeeze(ops.get_collection("IS_TRAINING"))
-        mean, variance = control_flow_ops.cond(
-            is_training, update_mean_var, lambda: (ema_mean, ema_var))
+        mean, variance = control_flow_ops.cond( is_training, update_mean_var,
```
Comment (ilblackdragon) [inline tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py]: extra space before `is_training`

**Comment by ilblackdragon** [comment]
@tensorflow-jenkins test this please.

**Comment by ilblackdragon** [comment]
@mikowals can you please resolve conflicts?

**Comment by mikowals** [comment]
I removed the spaces and squashed the commits to one.

**Comment by terrytangyuan** [comment]
@mikowals There's still conflict

**Comment by vrv** [comment]
status?

**Comment by ilblackdragon** [comment]
@tensorflow-jenkins test this please.

**Comment by vrv** [comment]
status?

**Comment by mikowals** [comment]
I do not understand the error messages coming from the Linux GPU PIP tests.  Is it possible they are linter related? 

I notice that lines 60 and 61 that I have written are into the right margin compared to the rest of the file and that since this pull originated many other lines in the file have had returns added with a commit message about fixing lint errors.

**Comment by vrv** [comment]
not sure, might be test infra failure.  if you rebase to resolve the conflict we can try again

**Comment by mikowals** [comment]
rebased to resolve confict.

**Comment by vrv** [comment]
Thanks!

@tensorflow-jenkins test this please

**Comment by mikowals** [comment]
Since this has yet to be merged I wonder if I should amend to it use the newly added  `tf.contrib.layers.batch_norm` to avoid duplication.

I think this could become:

    def batch_normalize(tensor_in,
           epsilon=1e-5,
           convnet=False, # this becomes redundant with tensor_in shape
           decay=0.9,
           scale_after_normalization=True):

        with vs.variable_scope("batch_norm"):
            batch_norm = functools.partial(layers.batch_norm, tensor_in,
                  decay=decay,
                  center=True,
                  scale=scale_after_normalization,
                  epsilon=epsilon,
                  updates_collections=None, 
                  scope='learn_bn')

            batch_norm_from_moving_averages = functools.partial(batch_norm, 
                  is_training=False, reuse=True)

            is_training = array_ops_.squeeze(ops.get_collection("IS_TRAINING"))
            return control_flow_ops.cond(is_training, batch_norm, batch_norm_from_moving_averages)

**Comment by vrv** [comment]
Sorry @mikowals, this PR dropped off the attention list.  @ilblackdragon what do you think about his suggestion?

**Comment by mikowals** [comment]
I have rebased to current master.  Sorry for close / reopen activity.  I am learning the limits of my git understanding.

File: tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py [hunk 3]
```diff
@@ -54,22 +54,31 @@ def batch_normalize(tensor_in,
...
+      update_moving_var = moving_averages.assign_moving_average(
+          moving_var, var, decay)
+      with ops.control_dependencies([update_moving_mean, update_moving_var]):
+        return array_ops_.identity(mean), array_ops_.identity(var)
 
     is_training = array_ops_.squeeze(ops.get_collection("IS_TRAINING"))
     mean, variance = control_flow_ops.cond(is_training, _update_mean_var,
-                                           lambda: (ema_mean, ema_var))
+                                           lambda: (moving_mean, moving_var))
     return nn.batch_norm_with_global_normalization(
```
Comment (fwalch) [inline tensorflow/contrib/learn/python/learn/ops/batch_norm_ops.py]: Since you're already touching this function, you could change this to `nn.batch_normalization` (`nn.batch_norm_with_global_normalization` is deprecated).

**Comment by ilblackdragon** [comment]
@tensorflow-jenkins test this please!

@mikowals Let's submit what you have (sorry for late responses on my side) and we can refactor to use `layers.batch_norm` in subsequent changes. Note, we ideally want to move `learn/ops/` into `layers/`, so it will be a bit more work.
